piptools/build.py

```diff
@@ -193,6 +201,15 @@ def _create_project_builder(
         yield build.ProjectBuilder(src_dir, runner=runner)
         return
 
+    if upgrade_packages is not None:
+        # Write packages to upgrade to a temporary file to set as
+        # constraints for the installation to the builder environment,
+        # in case build requirements are among them
+        tmpfile = tempfile.NamedTemporaryFile(mode="w+t", delete=False)
+        tmpfile.write("\n".join(package for package in upgrade_packages))
+        tmpfile.flush()
+        os.environ["PIP_CONSTRAINT"] = tmpfile.name
```

Comment:
So apparently this is not being cleaned up: neither the env file, nor the file descriptor behind `tempfile.NamedTemporaryFile()` is closed.

This was mentioned as necessary in https://github.com/jazzband/pip-tools/pull/1681#discussion_r1668032908.

The proper way to handle it would be to use a context manager.

Reply:
FTR, I'm taking over this. I moved it into another CM locally. Will push shortly.